experimental/algorithm/LAGraph_RPQMatrix.c

```diff
@@ -0,0 +1,197 @@
+
+    LG_ASSERT(lhs != NULL, GrB_NULL_POINTER);
+    LG_ASSERT(rhs != NULL, GrB_NULL_POINTER);
+
+    OK(LAGraph_RpqMatrix(lhs, msg));
+    OK(LAGraph_RpqMatrix(rhs, msg));
+
+    GrB_Matrix lhs_mat = lhs->res_mat;
+    GrB_Matrix rhs_mat = rhs->res_mat;
+
+    GRB_TRY(GrB_eWiseAdd(plan->res_mat, GrB_NULL, GrB_NULL,
+                         op, lhs_mat, rhs_mat, GrB_DESC_R));
+
+    return (GrB_SUCCESS);
+}
+
+static GrB_Info LAGraph_RpqMatrixConcat(RpqMatrixPlan *plan, char *msg)
+{
+    LG_ASSERT(plan != NULL, GrB_NULL_POINTER);
+    LG_ASSERT(plan->op == RPQ_MATRIX_OP_CONCAT, GrB_INVALID_VALUE);
+    LG_ASSERT(plan->res_mat == NULL, GrB_INVALID_VALUE);
+
+    RpqMatrixPlan *lhs = plan->lhs;
+    RpqMatrixPlan *rhs = plan->rhs;
+
+    LG_ASSERT(lhs != NULL, GrB_NULL_POINTER);
+    LG_ASSERT(rhs != NULL, GrB_NULL_POINTER);
+
+    OK(LAGraph_RpqMatrix(lhs, msg));
+    OK(LAGraph_RpqMatrix(rhs, msg));
+
+    GrB_Matrix lhs_mat = lhs->res_mat;
+    GrB_Matrix rhs_mat = rhs->res_mat;
+
+    GRB_TRY(GrB_mxm(plan->res_mat, GrB_NULL, GrB_NULL,
+                    sr, lhs_mat, rhs_mat, GrB_DESC_R));
+
+    return (GrB_SUCCESS);
+}
+
+static GrB_Info LAGraph_RpqMatrixKleene(RpqMatrixPlan *plan, char *msg)
+{
+    LG_ASSERT(plan != NULL, GrB_NULL_POINTER);
+    LG_ASSERT(plan->op == RPQ_MATRIX_OP_KLEENE, GrB_INVALID_VALUE);
+    LG_ASSERT(plan->res_mat == NULL, GrB_INVALID_VALUE);
+
+    RpqMatrixPlan *lhs = plan->lhs;
+    RpqMatrixPlan *rhs = plan->rhs;
+
+    // Kleene star should have one child. Always right.
+    LG_ASSERT(lhs == NULL, GrB_NULL_POINTER);
+    LG_ASSERT(rhs != NULL, GrB_NULL_POINTER);
+
+    OK(LAGraph_RpqMatrix(rhs, msg));
+
+    GrB_Matrix B = rhs->res_mat;
+
+    // Creating identity matrix.
+    GrB_Index n;
+    GRB_TRY(GrB_Matrix_nrows(&n, B));
+    GrB_Matrix E;
+    GRB_TRY(GrB_Matrix_new(&E, GrB_BOOL, n, n));
+
+    GrB_Vector v;
+    GRB_TRY(GrB_Vector_new(&v, GrB_BOOL, n));
+    GRB_TRY(GrB_Vector_assign_BOOL(v, NULL, NULL, true, GrB_ALL, n, NULL));
+
+    GRB_TRY(GrB_Matrix_diag(&E, v, 0));
+
+    GRB_TRY(GrB_Vector_free(&v));
+
+    // B + E
+    GrB_Matrix BPE;
+    GRB_TRY(GrB_Matrix_new(&BPE, GrB_BOOL, n, n));
+    GRB_TRY(GrB_eWiseAdd(BPE, GrB_NULL, GrB_NULL,
+                         op, B, E, GrB_DESC_R));
+    // S <- S x (B + E)
+    GrB_Matrix S, T;
+    GRB_TRY(GrB_Matrix_dup(&S, E));
+
+    bool changed = true;
+    GrB_Index nnz_S = n, nnz_T = 0;
+    while (changed)
+    {
+        // T = S * (B + E)
+        GRB_TRY(GrB_Matrix_new(&T, GrB_BOOL, n, n));
+        GRB_TRY(GrB_mxm(T, GrB_NULL, GrB_NULL,
+                    sr, S, BPE, GrB_DESC_R));
+
+        GRB_TRY(GrB_Matrix_nvals(&nnz_T, T));
+        if (nnz_T != nnz_S)
+        {
+            changed = true;
+            nnz_S = nnz_T;
+            GRB_TRY(GrB_Matrix_free(&S));
+            S = T;
+        }
+        else
+        {
+            changed = false;
+            GRB_TRY(GrB_Matrix_free(&T));
+        }
+    }
```

Comment:
Let's try fast exponentiation here.

```suggestion
    // S <- B + E
    GrB_Matrix S;
    GRB_TRY(GrB_Matrix_new(&S, GrB_BOOL, n, n));
    GRB_TRY(GrB_eWiseAdd(S, GrB_NULL, GrB_NULL,
                         op, B, E, GrB_DESC_R));

    // TODO: Take real nnz here. I think it would work but feels cursed.
    GrB_Index nnz_S = n, nnz_S_old = 0;
    while (nnz_S != nnz_S_old)
    {
        // S <- S x S
        GRB_TRY(GrB_mxm(S, GrB_NULL, GrB_NULL,
                    sr, S, S, GrB_NULL));

        nnz_S_old = nnz_S;
        GRB_TRY(GrB_Matrix_nvals(&nnz_S, S));
    }
```

Basically, this would calculate `(((((B + E)^2)^2)^2...)^2 = (B + E)^(2^n) = S` for some `n`. Assume `m = 2^n`. Then `S = B^m + B^(m - 1) + ... + E`.

Probably, it can be slower in GrB but it is worth a try. Nevertheless, the existing code might be rewritten in that way we use less matrices'n'replaces by employing the fact GraphBLAS does not replace the non-presented entries by default.